### PR TITLE
TLS CA Certs

### PR DESCRIPTION
**Pull request by sthulb**
I've added the ability to present the CA cert with the server certificates, this allow for greater trust between the server and client.

### CODE REVIEW

**Comment by stevvooe** [comment]
@sthulb How is this different from the existing functionality? Could you make sure to modify the documentation accordingly?

**Comment by sthulb** [comment]
@stevvooe this would be like `SSLCertificateChainFile` in Apache.

**Comment by stevvooe** [comment]
@sthulb How is that different from the existing functionality in the registry? The code seems identical to that of `ClientCAs`.

**Comment by sthulb** [comment]
@stevvooe the new code creates a pool for `pool.RootCAs`

**Comment by stevvooe** [comment]
@sthulb According to http://golang.org/pkg/crypto/tls/#Config, this removes the need for a client to provide the entire certificate chain. The configuration value should match the name `RootCAs` and the description of what it does should be added to docs/configuration.md.

**Comment by sthulb** [comment]
Yep. It allows the server to verify the server's cert for the client. I'll add the extra docs.

File: configuration/configuration_test.go [hunk 1]
```diff
@@ -64,18 +64,20 @@ var configStruct = Configuration{
...
 			Addr string `yaml:"addr,omitempty"`
 		} `yaml:"debug,omitempty"`
 	}{
 		TLS: struct {
-			Certificate string   `yaml:"certificate,omitempty"`
-			Key         string   `yaml:"key,omitempty"`
-			ClientCAs   []string `yaml:"clientcas,omitempty"`
+			Certificate    string   `yaml:"certificate,omitempty"`
+			Key            string   `yaml:"key,omitempty"`
+			CACertificates []string `yaml:"cacertificates,omitempty"`
```
Comment (stevvooe) [inline configuration/configuration_test.go]: To be clear, this should be called `RootCAs` to match up with the `crypto/tls` config.

**Comment by sthulb** [comment]
@stevvooe updated.

File: configuration/configuration.go [hunk 2]
```diff
@@ -74,6 +74,10 @@ type Configuration struct {
 			// Certificate.
 			Key string `yaml:"key,omitempty"`
 
+			// Certificate Authority certificates for clients to verify
```
Comment (stevvooe) [inline configuration/configuration.go]: Please follow Go's documentation guidelines. This should start with "RootCAs".

File: docs/configuration.md [hunk 3]
```diff
@@ -886,6 +892,17 @@ and proxy connections to the registry server.
...
   </tr>
+  <tr>
+    <td>
+      <code>rootcas</code>
+    </td>
+    <td>
+      yes
+    </td>
+    <td>
+      An array of paths to x509 CA certificate files.
```
Comment (stevvooe) [inline docs/configuration.md]: Please clarify that this is for the client certificate chain.

**Comment by sthulb** [comment]
@stevvooe Fixed both points, thanks

File: docs/configuration.md [hunk 4]
```diff
@@ -886,6 +892,17 @@ and proxy connections to the registry server.
...
   </tr>
+  <tr>
+    <td>
+      <code>rootcas</code>
+    </td>
+    <td>
+      yes
+    </td>
+    <td>
+      An array of paths to the root CA files, as x509.
```
Comment (stevvooe) [inline docs/configuration.md]: "An array of paths to root CA files in x509 format used to verify client certificates."

Comment (sthulb) [inline docs/configuration.md]: This change has nothing to do with client certs. It's to prove validation of the serverside certs

Comment (stevvooe) [inline docs/configuration.md]: Great. So, could you clarify it? Please see Go's documentation for clarification:

```
// RootCAs defines the set of root certificate authorities
// that clients use when verifying server certificates.
// If RootCAs is nil, TLS uses the host's root CA set.
```

If the documentation is not clear, the feature will not be added.

**Comment by stevvooe** [comment]
@sthulb Could you please rebase this? I'd like to merge it for 2.1.

cc @diogomonica @NathanMcCauley PTAL

File: cmd/registry/main.go [hunk 5]
```diff
@@ -77,22 +77,18 @@ func main() {
...
-			for _, ca := range config.HTTP.TLS.ClientCAs {
-				caPem, err := ioutil.ReadFile(ca)
-				if err != nil {
-					context.GetLogger(app).Fatalln(err)
-				}
-
-				if ok := pool.AppendCertsFromPEM(caPem); !ok {
-					context.GetLogger(app).Fatalln(fmt.Errorf("Could not add CA to pool"))
-				}
+		if len(config.HTTP.TLS.RootCAs) != 0 {
```
Comment (diogomonica) [inline cmd/registry/main.go]: Aren't these supposed to be ClientCAs? Calling them RootCAs might be confusing.

File: cmd/registry/main.go [hunk 6]
```diff
@@ -262,3 +258,22 @@ func debugServer(addr string) {
 		log.Fatalf("error listening on debug interface: %v", err)
 	}
 }
+
+type reader func(path string) ([]byte, error)
+
+func createCertPool(certs []string, fileReader reader) (*x509.CertPool, error) {
```
Comment (diogomonica) [inline cmd/registry/main.go]: Can you give me an usecase where we would use a different reader other than `ioutil.ReadFile`?

Comment (diogomonica) [inline cmd/registry/main.go]: Is this purely for testability?

Comment (stevvooe) [inline cmd/registry/main.go]: Yes, appears so.

**Comment by diogomonica** [comment]
I don't get why this functionality is needed.

@sthulb even SSLCertificateChainFile is being deprecated in liu of adding the chain to SSLCertificateFile: http://www.bensprotips.com/2014/04/apache-sslcertificatechainfile-directive-deprecated-sslcertificatefile/

If we need to serve intermediates we should be hosting them in the same convention that web servers follow (host the chain in one file).

**Comment by sthulb** [comment]
@diogomonica @stevvooe Hi, sorry for the delay... This PR was to just allow users to present their CA cert with the server certificate for extra validation. It also gives the user the ability to trust more than just their registry, but their entire CA chain.

Whilst this is similar to `SSLCertficateChainFile`, that's due to Go only supporting that as an option. We could artificially limit it to a cert in a file.

**Comment by diogomonica** [comment]
@sthulb sorry, still not getting the use-case. Can you make it a bit more explicit?

> Allow users to present their CA cert with the server certificate for extra validation

What are "users" in this sentence. Does this mean: including the whole CA chain in the certificate served to the clients connecting to the registry? And if so, can't we just include the whole chain in the server.pem?

> It also gives the user the ability to trust more than just their registry, but their entire CA chain.

How does the server configuration change in regards to this. This is a matter of the clients trusting a CA instead of  the server certificate directly, no?

**Pull request closed without merging**